packages/winpanda/extra/src/winpanda/core/rc_ctx.py

```diff
@@ -0,0 +1,207 @@
+"""Panda package management for Windows.
+
+Resource rendering context calculation stuff.
+"""
+import configparser as cfp
+import json
+
+from common import constants as cm_const
+from common import logger
+from common.storage import ISTOR_NODE, IStorNodes
+from core.package.id import PackageId
+
+
+LOG = logger.get_logger(__name__)
+
+
+class RCCONTEXT_ITEM:
+    """Element of resource rendering context."""
+    MASTER_PRIV_IPADDR = 'master_priv_ipaddr'
+    LOCAL_PRIV_IPADDR = 'local_priv_ipaddr'
+    ZK_CLIENT_PORT = 'zk_client_port'
+
+    DCOS_INST_DPATH = 'dcos_inst_dpath'
+    DCOS_CFG_DPATH = 'dcos_cfg_dpath'
+    DCOS_TMP_DPATH = 'dcos_tmp_dpath'
+
+    PKG_INST_DPATH = 'pkg_inst_dpath'
+    PKG_LOG_DPATH = 'pkg_log_dpath'
+    PKG_RTD_DPATH = 'pkg_rtd_dpath'
+    PKG_WORK_DPATH = 'pkg_work_dpath'
+    PKG_SHRCFG_DPATH = 'pkg_shrcfg_dpath'
+
+
+class ResourceContext:
+    """Resource rendering context manager."""
+    def __init__(self, istor_nodes=None, cluster_conf=None, pkg_id=None):
+        """Constructor.
+
+        :param istor_nodes:  IStorNodes, DC/OS installation storage nodes (set
+                             of pathlib.Path objects)
+        :param cluster_conf: dict, configparser.ConfigParser.read_dict()
+                             compatible data. DC/OS cluster setup parameters
+        :param pkg_id:       PackageId, package ID
+        """
+        if istor_nodes is not None:
+            assert isinstance(istor_nodes, IStorNodes), (
+                f'Argument: istor_nodes:'
+                f' Got {type(istor_nodes).__name__} instead of IStorNodes'
+            )
+        if cluster_conf is not None:
+            assert isinstance(cluster_conf, dict), (
+                f'Argument: cluster_conf:'
+                f'Got {type(cluster_conf).__name__} instead of dict'
+            )
+        if pkg_id is not None:
+            assert isinstance(pkg_id, PackageId), (
+                f'Argument: pkg_id: PackageId is required: {pkg_id}'
+            )
+
+        self._istor_nodes = istor_nodes
+        self._cluster_conf = cluster_conf
+        self._pkg_id = pkg_id
+
+    def get_items(self, json_ready=False):
+        """Get resource rendering context items.
+
+        :param json_ready: bool, get JSON-compatible context items, if True
+        :return:           dict, set of resource rendering context items
+        """
+        retrievers = (self._get_istor_items,
+                      self._get_cluster_conf_items,
+                      self._get_pkg_items)
+        items = {}
+
+        for retriever in retrievers:
+            items.update(retriever(json_ready))
+
+        return items
+
+    def _get_istor_items(self, json_ready=False):
+        """Discover resource rendering context items from DC/OS installation
+        storage configuration.
+
+        :param json_ready: bool, get JSON-compatible context items, if True
```

Comment:
Why do we need this? Is stripping double-quotes the only extra processing needed? e.g. could there be escapes in the string?

Reply:
Not exactly... When we do preliminary Jinja rendering for a JSON-formatted template we can face with the situation when our variable for substitution contains backslahes. So after rendering that template would contain character sequences wrongly interpreted by JSON-parser as escape sequences and causing parsing errors. This is the exact case for substituting variable parts of any FS path on windows. So, to handle such cases we have to screen every single back slash in such variables before doing rendering. I suppose that the most correct approach to give this work to the standard JSON serializer. And ater the JSON0-serializer does all the escaping in variables the only thing laft is to strip double quotes.